### ISSUE TITLE
ERROR: modpost: "__mulodi4" [drivers/block/nbd.ko] undefined!

### ISSUE THREAD

**Issue by nathanchance** [ClangBuiltLinux/linux#1438]
After commit [fad7cd3310db](https://git.kernel.org/axboe/linux-block/c/fad7cd3310db3099f95dd34312c77740fbc455e5) ("nbd: add the check to prevent overflow in __nbd_ioctl()") in the block tree, the following error is seen with several different 32-bit configurations:

```
$ make -skj"$(nproc)" ARCH=arm CROSS_COMPILE=arm-linux-gnueabi- LLVM=1 LLVM_IAS=1 distclean allmodconfig all
...
ERROR: modpost: "__mulodi4" [drivers/block/nbd.ko] undefined!
...
```

`check_mul_overflow()` will ultimately call `__builtin_mul_overflow()`, where I assume that it is problematic that the operands are `long long` so LLVM calls to the compiler-rt functions that we do not link against.

cc @kees

**Comment by nickdesaulniers** [comment]
there's an upstream bug for this (@stephenhines and @kraj are on it): https://llvm.org/pr28629

**Comment by stephenhines** [comment]
The bad news is that this is probably still very low priority to get fixed. For Android, we just started out by including sub-components of compiler-rt directly with libgcc, and later switched completely to Clang's own builtin library with no libgcc.

**Comment by rengolin** [comment]
My attempts to reproduce a long long multiplication come out lowered from clang on all Arm 32/64 soft/hard I could come up with, with the only exception being `-march=armv6 -mthumb`, which calls `__aeabi_lmul`, not `__mulodi4`. It's possible they're aliases and that's what the symbol points to? Is this it?

But when I try to link, it just works. clang is calling `arm-linux-gnueabi-ld`. I'm not trying freestanding, so that may be another reason I can't reproduce?

I think the "low priority" of it is because of the fuzzy nature of the original reports, which had `__has_builtin`, sanitizers and all sorts of other red herrings that it took too long to narrow down. I don't think this is a "low priority" bug per se, but it seems that it's such a corner case that was always worked around, so no one really bothered to go down the rabbit hole.

With a small enough reproducer, it should be much clearer to find where the problem is and perhaps it's even easy to implement. I'm guessing the pattern is not enabled in Thumb and ends up going to the runtime. It could be a simple case of adding a custom lowering for Thumb...

**Comment by nathanchance** [comment]
I'll see if I can get a smaller reproducer tomorrow then.

NOTE: This is not just a 32-bit ARM error, I do also see it on 32-bit MIPS.

**Comment by nickdesaulniers** [comment]
We can probably creduce the kernel easily.

**Comment by rengolin** [comment]
Thanks! Please attach the reproducer to the original bug, so other people can see it, and we update this if/once we fix it.

**Comment by rengolin** [comment]
> NOTE: This is not just a 32-bit ARM error, I do also see it on 32-bit MIPS.

This sounds like people just didn't get to it because no one cared enough about 32-bits by the time the change was introduced...

**Comment by nathanchance** [comment]
Reproducer posted: https://llvm.org/pr28629

**Comment by nickdesaulniers** [comment]
If we can fix clang for ARM and MIPS at least, then we might be able to ship in the kernel:
```diff
diff --git a/drivers/block/nbd.c b/drivers/block/nbd.c
index 5170a630778d..2271a50ba4c0 100644
--- a/drivers/block/nbd.c
+++ b/drivers/block/nbd.c
@@ -1388,6 +1388,12 @@ static void nbd_set_cmd_timeout(struct nbd_device *nbd, u64 timeout)
                blk_queue_rq_timeout(nbd->disk->queue, 30 * HZ);
 }
 
+// FIXME: https://llvm.org/pr28629
+// https://github.com/ClangBuiltLinux/linux/issues/1438
+#if defined(__clang__) && (defined(__arm__) || (defined(__mips__))
+#define MULODI4_BROKEN
+#endif
+
 /* Must be called with config_lock held */
 static int __nbd_ioctl(struct block_device *bdev, struct nbd_device *nbd,
                       unsigned int cmd, unsigned long arg)
@@ -1408,8 +1414,10 @@ static int __nbd_ioctl(struct block_device *bdev, struct nbd_device *nbd,
        case NBD_SET_SIZE:
                return nbd_set_size(nbd, arg, config->blksize);
        case NBD_SET_SIZE_BLOCKS:
+#ifndef MULODI4_BROKEN
                if (check_mul_overflow((loff_t)arg, config->blksize, &bytesize))
                        return -EINVAL;
+#endif
                return nbd_set_size(nbd, bytesize, config->blksize);
        case NBD_SET_TIMEOUT:
                nbd_set_cmd_timeout(nbd, arg);
```
but with an added version check. Err, that would use `bytesize` uninitialized. We'd maybe go back to using `arg * config->blksize` instead.  Errr...the kernel patch that exposed this fixes a bug; we'd probably wan't to use a hand written `check_mul_overflow` that doesn't rely on a builtin.

**Comment by nathanchance** [comment]
Shouldn't we be working around this issue generically, rather than in that specific call site? Initial idea would be breaking up the `COMPILER_HAS_GENERIC_BUILTIN_OVERFLOW` define into add, sub, and mul then defining the mul one only if we know we are working with a good compiler and architecture combination. See `include/linux/overflow.h`.

**Comment by rengolin** [comment]
Quick update: This problem was mostly fixed earlier this month (by @topperc) for x86 and @nickdesaulniers is working to get that propagated to Arm and Mips. It will only fix from now on (clang 14+), leaving previous clangs with the same problem.

We may be able to get this in clang 13 if we can backport the current RC3. No promises, though, as it's very late in the release process. But if it works, we would get 13 with the fix in a a few weeks.

That still leaves all previous clangs (12 and older) unfixed (we don't backport beyond six months). How the kernel is going to fix for those older versions is up to you guys.

**Comment by nickdesaulniers** [comment]
Some toolchain patches have gone in:
32b ARM: https://reviews.llvm.org/rG5c91b98c5d45243352bf10262454bcac77cd3fed
32b MIPS 1: https://reviews.llvm.org/rGc8c176d999d2c2166dee8d67c32b4f8f42b4f1bd

MIPS is now hitting `__multi3` being undefined. @rengolin mentioned that perhaps the above 2 could be cherry picked to clang-13, but that would depend on https://reviews.llvm.org/rGb2ca4dc935859b324fd7e9ca804160913a7468a5 and @tstellar + @topperc 's feedback).  Since we need to work around this in the kernel anyways, I'm not interested in rushing this, unless the clang-13 release is perhaps still "far" away (like a month).

**Comment by topperc** [comment]
Is 32-bit X86 still broken for __mulodi4?

**Comment by rengolin** [comment]
It seems so. In your [original patch](https://reviews.llvm.org/D107581) you only add it for 128-bits, not 64-bits. However, further up the X86Lowering code, `ISD::SMULO` is legal (custom lowered) for `MVT::i64` and it doesn't seem to be conditional on 32/64 bits.

But when I compile the original code with `-target i386` I still get the `__mulodi4` call. Perhaps the X86 back-end needs the same handling for `RTLIB` on 64bits?

**Comment by topperc** [comment]
> It seems so. In your [original patch](https://reviews.llvm.org/D107581) you only add it for 128-bits, not 64-bits. However, further up the X86Lowering code, `ISD::SMULO` is legal (custom lowered) for `MVT::i64` and it doesn't seem to be conditional on 32/64 bits.
> 
> But when I compile the original code with `-target i386` I still get the `__mulodi4` call. Perhaps the X86 back-end needs the same handling for `RTLIB` on 64bits?

Isn't it in this for loop that has a continue for i64 on 32-bit targets?

```
  for (auto VT : { MVT::i8, MVT::i16, MVT::i32, MVT::i64 }) {                                                                                                                                                                                                
    if (VT == MVT::i64 && !Subtarget.is64Bit())                                                                                                                                                                                                              
      continue;                                                                                                                                                                                                                                              
    // Add/Sub/Mul with overflow operations are custom lowered.                                                                                                                                                                                              
    setOperationAction(ISD::SADDO, VT, Custom);                                                                                                                                                                                                              
    setOperationAction(ISD::UADDO, VT, Custom);                                                                                                                                                                                                              
    setOperationAction(ISD::SSUBO, VT, Custom);                                                                                                                                                                                                              
    setOperationAction(ISD::USUBO, VT, Custom);                                                                                                                                                                                                              
    setOperationAction(ISD::SMULO, VT, Custom);                                                                                                                                                                                                              
    setOperationAction(ISD::UMULO, VT, Custom);      
```

**Comment by rengolin** [comment]
> Isn't it in this for loop that has a continue for i64 on 32-bit targets?

aha! Yes, I missed that. Been looking at this from different computers and mobiles, lacking context on all of them.

Not sure the kernel cares about i386 anymore but with your legalisation fix, it seems the same thing would work just like Arm and Mips.

**Comment by topperc** [comment]
> > Isn't it in this for loop that has a continue for i64 on 32-bit targets?
> 
> aha! Yes, I missed that. Been looking at this from different computers and mobiles, lacking context on all of them.
> 
> Not sure the kernel cares about i386 anymore but with your legalisation fix, it seems the same thing would work just like Arm and Mips.

Are we heading in the direction of removing MULO lib calls completely from llvm? My initial patch was just to fix the cases that wouldn’t link with either libgcc or compiler-rt.

**Comment by rengolin** [comment]
> Are we heading in the direction of removing MULO lib calls completely from llvm? My initial patch was just to fix the cases that wouldn’t link with either libgcc or compiler-rt.

Maybe? I'm not advocating for it, I'm just trying to fix an issue that exists for over 5 years and still doesn't have a nice solution.

Linking compiler-rt isn't mandatory and not used on many systems. Linking with compiler-rt and libgcc won't work, so if there is functionality that is only available in compiler-rt, we can't know if the linker will find it or not, so lowering calls to `__mulodi4` is bound to fail on many systems.

The current legalisation is un-optimised, but Compiler-RT's implementation is plain-C for all platforms, so not the most optimised either. If we can lower the builtin with the compiler and avoid the linking errors on anything that doesn't use compiler-rt, then I think it's a good compromise.

Even if libgcc implements it, the kernel and embedded software won't have it, so are bound to have the same problem. That's why they use `__has_builtin` all over the place and if we say we have it, then we better implement it. But we don't, we belay that for compiler-rt, which they don't link against.

I don't think eliding the calls as a target option is the best option, but right now it's the only option. I don't really know how Clang can emit code that it knows compiler-rt implements when it doesn't know if it will be linked against it.

There was a discussion many years ago to split builtins from compiler-rt and make it so clang *always* linked against that. It would obviously clash with symbols from libgcc and other RT libs, but we could mark them weak or some other linkage magic that escapes me now.

Perhaps we need some more thorough discussion on the list...

**Comment by kraj** [comment]
for compatibility reasons I think its best to remove it from clang.

**Comment by nickdesaulniers** [comment]
> But when I compile the original code with -target i386 I still get the __mulodi4 call. Perhaps the X86 back-end needs the same handling for RTLIB on 64bits?

> Not sure the kernel cares about i386 anymore but with your legalisation fix, it seems the same thing would work just like Arm and Mips.

Yes, this is broken for 32b x86 kernels as well (we just don't have CI coverage for ARCH=i386 defconfig + CONFIG_BLK_DEV_NBD=y). If I enable `CONFIG_BLK_DEV_NBD=y`, then I observe the following linkage failure:
```
ld.lld: error: undefined symbol: __mulodi4
```

**Comment by nickdesaulniers** [comment]
mips patch 2/2: https://reviews.llvm.org/D108926
i386 patch: https://reviews.llvm.org/D108928

For future travelers, llvm/include/llvm/IR/RuntimeLibcalls.def has the relevant mapping for these functions.

**Comment by nickdesaulniers** [comment]
ppc32: https://reviews.llvm.org/D108936
riscv: https://reviews.llvm.org/D108939; note: we don't have coverage of that target for kernel builds, so I can't quickly verify.

Clang is passed `--rtlib=compiler-rt` but I'll bet that's not propagated in the IR, so things like LTO might not know for ISel whether compiler-rt would be available or not.

**Comment by nickdesaulniers** [comment]
This is all fixed up on the LLVM side, see the six commits referenced from https://llvm.org/pr28629.

Now to fix this up on the kernel side. This is top priority for me since it's starting to affect Android builds now, too.
https://android-review.googlesource.com/c/kernel/common/+/1820696

**Comment by nickdesaulniers** [comment]
boy, this is going to be tricky. I have a kernel patch that detects when the compiler would generate a libcall to __mulodi4, and instead falls back to the non `COMPILER_HAS_GENERIC_BUILTIN_OVERFLOW` implementation of `check_mul_overflow`: https://github.com/ClangBuiltLinux/linux/blob/83d09ad4b950651a95d37697f1493c00d888d0db/include/linux/overflow.h#L118-L201

The problem is that the implementation of `check_mul_overflow` that doesn't use `__builtin_mul_overflow` has a division that results in a linkage failure against `__aeabi_uldivmod`. The kernel defines `__aeabi_uidivmod`, but not `__aeabi_uldivmod`. I don't think we can use `do_div` macro in `check_mul_overflow` since `check_mul_overflow` is more agnostic to the types of the parameters, since `__builtin_mul_overflow` is rather type flexible.

Also, I need to file a bug against libgcc about implementing __mulodi4.

**Comment by nickdesaulniers** [comment]
Specifically `__aeabi_uldivmod` is for ARCH=arm. I can't fall back to the non-`COMPILER_HAS_GENERIC_BUILTIN_OVERFLOW` implementations of `check_mul_overflow` anyways; they have a multiplication that produces `__mulodi4` on ARCH=i386 anyways...

**Comment by rengolin** [comment]
According to the ARM RTABI document, `_aeabi_uldivmod` is defined as a runtime helper (paragraph 4.2), unlike the overflow variants.

Both libgcc and compiler-rt implement that, but the kernel doesn't use either. If you want to make use of it you have to either include one of the run-time libraries or add it with the others the kernel implements. Being the kernel, I believe the only solution is the latter. But you'll have to do that for all relevant 32-bit platforms again.

[This](https://stackoverflow.com/questions/34627213/undefined-reference-to-aeabi-ldivmod-when-building-kernel-for-arm32) is 5 years old and looks very similar to what you're doing.

**Comment by nickdesaulniers** [comment]
I feel like we'll either need to add an implementation of `__mulodi4` to the kernel; either that or that plus my messy patch AND `__aeabi_uldivmod`, so adding `__mulodi4` is probably less work.

**Comment by nickdesaulniers** [comment]
god damn...I re-implemented `__mulodi4` from compiler-rt in the kernel...it depends on `__aeabi_uldivmod`...argh!!!!

**Comment by nathanchance** [comment]
The 64-bit division in the `!COMPLER_HAS_GENERIC_BUILTIN_OVERFLOW` case has implications for older GCCs too: https://lore.kernel.org/r/20210909182525.372ee687@canb.auug.org.au/

cc @kees

@nickdesaulniers I assume there is division in there. Could you use the 32-bit division helpers to fix that error?

**Comment by nickdesaulniers** [comment]
> Could you use the 32-bit division helpers to fix that error?

Yes! Thank you! (https://stackoverflow.com/a/52038732)

> The 64-bit division in the !COMPLER_HAS_GENERIC_BUILTIN_OVERFLOW case has implications for older GCCs too: https://lore.kernel.org/r/20210909182525.372ee687@canb.auug.org.au/

Oh, that was literally reported today!

**Comment by nickdesaulniers** [comment]
ok, thinking about this more over a cup of coffee and a walk:
1. https://lore.kernel.org/r/20210909182525.372ee687@canb.auug.org.au/ needs to be fixed via do_div or relevant helper from linux/math64.h such as `div64_s64`, regardless of this issue, though it will help.
2. skip providing `__mulodi4`; instead break up `COMPLER_HAS_GENERIC_BUILTIN_OVERFLOW` as per Nathan since technically clang did not have `__builtin_mul_overflow` that didn't require linking against compiler-rt for 32b targets.
3. disable the `COMPLER_HAS_GENERIC_BUILTIN_OVERFLOW` for multiply for older clang revisions.

**Comment by Villemoes** [comment]
Sorry I don't have time to do any actual work. Random thoughts on $subject:

We can certainly patch the  fallbacks to use whatever variant of do_div is appropriate. But AFAIK, for gcc, they are only used for gcc 4.9; gcc 5.1+ has the builtin_overflow checkers. I'd much rather push for bumping the required gcc version to 5.1 and get rid of all my nasty fallback macros. 

As for clang, are there any supported versions of clang which do not claim 
```
 __has_builtin(__builtin_mul_overflow) && \
  __has_builtin(__builtin_add_overflow) && \
  __has_builtin(__builtin_sub_overflow)
```
?

Per above, I'd really like to get rid of the fallbacks, and once the required gcc version gets bumped to 5.1 (which will happen sooner or later for some other reasons anyway), it would be nice if we didn't have to keep the multiply fallback around because of clang. Is there any problem (licensing???) including whatever support routines clang might emit calls to in the kernel? And gosh, why do those libgcc/compiler-rt functions need to have such unreadable and unintuitive names.

Oh, and whatever fixes are made, please include a patch to remove the `#if BITS_PER_LONG == 64` checks from `lib/test_overflow.c` so that we have a simple way to ensure those fixes get build and runtime tested.

**Comment by rengolin** [comment]
Some random comments...

> I'd much rather push for bumping the required gcc version to 5.1 and get rid of all my nasty fallback macros.

GCC 5.1 is rather old, so ok to bump requirements to it. But we're fixing these things in Clang now, so will be ready on the next version (end-2021).

> As for clang, are there any supported versions of clang which do not claim
> 
> ```
>  __has_builtin(__builtin_mul_overflow) && \
>   __has_builtin(__builtin_add_overflow) && \
>   __has_builtin(__builtin_sub_overflow)
> ```

Probably very old versions... `__mulodi4` was implemented in compiler-rt in 2011... I'm guessing clang hooks soon followed.

> Is there any problem (licensing???) including whatever support routines clang might emit calls to in the kernel?

All mono-repo LLVM components have the same license (including clang and compiler-rt) and that license is more permissive than GPL. IANAL, but it should be fine to include compiler-rt code anywhere (or at least that's the intention).

>  And gosh, why do those libgcc/compiler-rt functions need to have such unreadable and unintuitive names.

Blame GCC. :) We just copied the names for compatibility.

**Comment by nickdesaulniers** [comment]
> I'd much rather push for bumping the required gcc version to 5.1 and get rid of all my nasty fallback macros.

I agree, otherwise we need to add some even more crazy ones to handle 64b division on 32b hosts: https://gist.github.com/nickdesaulniers/2479818f4983bbf2d688cebbab435863 is not correct and already a mess.

**Comment by nickdesaulniers** [comment]
Sent: https://lore.kernel.org/lkml/20210910234047.1019925-1-ndesaulniers@google.com/

**Comment by nickdesaulniers** [comment]
Linus merged my GCC 5.1 version bump after tagging v5.15rc1: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=76ae847497bc5207c479de5e2ac487270008b19b

I have a patch anyways for stable using `__builtin_choose_expr` in ways that I should be put in prison.
https://lore.kernel.org/stable/20210913203201.1844253-1-ndesaulniers@google.com/

**Comment by nickdesaulniers** [comment]
Linus suggested another approach to the fix in mainline that might be a more suitable backport to stable.
https://lore.kernel.org/stable/CAHk-=whiQBofgis_rkniz8GBP9wZtSZdcDEffgSLO62BUGV3gg@mail.gmail.com/
rather than
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=fad7cd3310db3099f95dd34312c77740fbc455e5

**Comment by nickdesaulniers** [comment]
v3 kernel patch: https://lore.kernel.org/llvm/20210920232533.4092046-1-ndesaulniers@google.com/

**Comment by nickdesaulniers** [comment]
https://lore.kernel.org/lkml/20210920232533.4092046-1-ndesaulniers@google.com/is the latest kernel patch in this saga.

**Comment by nickdesaulniers** [comment]
accepted: https://git.kernel.org/pub/scm/linux/kernel/git/axboe/linux-block.git/commit/?h=block-5.15&id=41e76c6a3c83c85e849f10754b8632ea763d9be4

(once this lands in mainline, I'll need to chase this into stable, then re-enable CONFIG_BLK_DEV_NBD=y in ACK)

**Comment by nickdesaulniers** [comment]
41e76c6a3c83c85e849f10754b8632ea763d9be4 is in v5.15-rc4. GKH pulled this into stable 5.14.
